arangod/Transaction/Methods.cpp

```diff
@@ -1751,162 +2028,128 @@ Future<OperationResult> transaction::Methods::removeLocal(
+        value = keyBuilder->slice();
       }
     } else if (value.isObject()) {
       VPackSlice keySlice = value.get(StaticStrings::KeyString);
       if (!keySlice.isString()) {
-        return Result(TRI_ERROR_ARANGO_DOCUMENT_HANDLE_BAD);
+        return res.reset(TRI_ERROR_ARANGO_DOCUMENT_HANDLE_BAD);
       }
       key = keySlice.stringView();
-    } else {
-      return Result(TRI_ERROR_ARANGO_DOCUMENT_HANDLE_BAD);
     }
 
-    // Primary keys must not be empty
+    // primary key must not be empty
     if (key.empty()) {
-      return Result(TRI_ERROR_ARANGO_DOCUMENT_HANDLE_BAD);
+      return res.reset(TRI_ERROR_ARANGO_DOCUMENT_HANDLE_BAD);
     }
 
-    previous.clear();
+    std::pair<LocalDocumentId, RevisionId> lookupResult;
+    res = collection->getPhysical()->lookupKey(this, key, lookupResult,
+                                               ReadOwnWrites::yes);
+    if (res.fail()) {
+      return res.reset(TRI_ERROR_ARANGO_DOCUMENT_NOT_FOUND);
+    }
 
-    auto res = collection->remove(*this, value, options, previous);
+    TRI_ASSERT(lookupResult.first.isSet());
+    TRI_ASSERT(lookupResult.second.isSet());
+    LocalDocumentId oldDocumentId = lookupResult.first;
+    RevisionId oldRevisionId = lookupResult.second;
+
+    previousDocumentBuilder->clear();
+    res = collection->getPhysical()->lookupDocument(
+        *this, oldDocumentId, *previousDocumentBuilder,
+        /*readCache*/ true, /*fillCache*/ false, ReadOwnWrites::yes);
 
     if (res.fail()) {
-      if (res.is(TRI_ERROR_ARANGO_CONFLICT) && !isBabies) {
-        TRI_ASSERT(previous.revisionId().isSet());
-        buildDocumentIdentity(collection.get(), resultBuilder, cid, key,
-                              previous.revisionId(), RevisionId::none(),
-                              options.returnOld ? &previous : nullptr, nullptr);
-      }
       return res;
     }
 
-    if (!options.silent) {
-      TRI_ASSERT(!options.returnOld || !previous.empty());
-      TRI_ASSERT(previous.revisionId().isSet());
-      buildDocumentIdentity(collection.get(), resultBuilder, cid, key,
-                            previous.revisionId(), RevisionId::none(),
-                            options.returnOld ? &previous : nullptr, nullptr);
+    bool excludeFromReplication = excludeAllFromReplication;
+    TRI_ASSERT(previousDocumentBuilder->slice().isObject());
+
+    res = removeLocalHelper(*collection, value, oldDocumentId, oldRevisionId,
+                            previousDocumentBuilder->slice(), options);
+
+    if ((res.is(TRI_ERROR_ARANGO_CONFLICT) && !isArray) ||
+        (res.ok() && !options.silent)) {
+      TRI_ASSERT(oldRevisionId.isSet());
+      buildDocumentIdentity(
+          *collection, resultBuilder, cid, key, oldRevisionId,
+          RevisionId::none(),
+          options.returnOld ? previousDocumentBuilder.get() : nullptr, nullptr);
+    }
+
+    if (res.ok() && !excludeFromReplication) {
+      replicationData->openObject(true);
```

Comment:
Just for me: When I should to write true for openObject or openArray

Reply:
`VPackBuilder::openObject(/*unindexed*/ = false)`

_unindexed_ here means we will write a VPack type Object _without_ an index table. This can save space and processing time for building the index table. At this code location, we know that we will only write two keys into the object, so it is a small optimization I recycled from previous code.

Reply:
Thanks, so true can be useful as optimization when I know I always have only few values, or doesn't have access to them via get/at?

Reply:
Yes, exactly!